src/lib.rs

```diff
@@ -84,62 +92,57 @@ pub enum VersionMatch {
     pub fn matches(&self, requested: &RequestedVersion) -> VersionMatch {
+        use RequestedVersion::*;
+        use VersionMatch::*;
+
         match requested {
-            RequestedVersion::Any => VersionMatch::Loosely,
-            RequestedVersion::Loose(major) => if self.major == *major {
-                VersionMatch::Loosely
+            Any => Loosely,
+            Loose(major) => if self.major == *major {
+                Loosely
             } else {
-                VersionMatch::NotAtAll
+                NotAtAll
             },
-            RequestedVersion::Exact(major, minor) => {
+            Exact(major, minor) => {
                 if self.major == *major && self.minor == *minor {
-                    VersionMatch::Exactly
+                    Exactly
                 } else {
-                    VersionMatch::NotAtAll
+                    NotAtAll
                 }
             }
         }
     }
 }
 
 /// Converts a `Vec<char>` to a `VersionComponent` integer.
-fn char_vec_to_int(char_vec: &Vec<char>) -> Result<VersionComponent, String> {
-    let joined_string = char_vec.into_iter().collect::<String>();
-    let parse_result = joined_string.parse::<VersionComponent>();
-    parse_result.or(Err(format!(
-        "error converting {:?} to a number",
-        joined_string
-    )))
+fn char_vec_to_int(char_vec: &[char]) -> Result<VersionComponent, String> {
+    let joined_string = char_vec.iter().collect::<String>();
+    let parse_result = joined_string.parse();
+    parse_result.map_err(|_| format!("error converting {:?} to a number", joined_string))
 }
 
 /// Attempts to find a version specifier from a CLI argument.
 ///
 /// It is assumed that the flag from the command-line is passed as-is
 /// (i.e. the flag starts with `-`).
-pub fn version_from_flag(arg: &String) -> Option<RequestedVersion> {
-    assert!(arg.starts_with("-"));
-    let mut version = arg.clone();
-    version.remove(0);
-    match RequestedVersion::from_string(&version) {
-        Ok(version) => Some(version),
-        Err(_) => None,
-    }
+pub fn version_from_flag(arg: &str) -> Option<RequestedVersion> {
+    assert!(arg.starts_with('-'));
+    RequestedVersion::from_str(&arg[1..]).ok()
 }
 
 /// Returns the entries in `PATH`.
-pub fn path_entries() -> Vec<path::PathBuf> {
-    let path_val = match env::var_os("PATH") {
-        Some(val) => val,
-        None => return Vec::new(),
-    };
-    env::split_paths(&path_val).collect()
+pub fn path_entries() -> Vec<PathBuf> {
+    if let Some(path_val) = env::var_os("PATH") {
+        env::split_paths(&path_val).collect()
+    } else {
+        Vec::new()
+    }
 }
 
 /// Gets the contents of a directory.
 ///
 /// Exists primarily to unwrap and ignore any unencodeable names.
-pub fn directory_contents(path: &path::PathBuf) -> collections::HashSet<path::PathBuf> {
-    let mut paths = collections::HashSet::new();
+pub fn directory_contents(path: &Path) -> HashSet<PathBuf> {
```

Comment:
I take it this works because `Path` isn't concrete like `PathBuf` so you're able to say you want the `Path` interface while still returning the concrete `PathBuf` instance? IOW `Path` not implementing `Sized` is fine since it isn't critical to the compiler to know the size of what is being returned?

Reply:
`Path` is to `str` as `PathBuf` is to `String`. One doesn't exist concretely, the other is owned.